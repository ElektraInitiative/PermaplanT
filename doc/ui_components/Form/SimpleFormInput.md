# SimpleFormInput

## Description

Textfield that can be used for single line or multi-line input.

## Props

| Name          | Type                   | Optional | Description                                                                                                       |
| ------------- | ---------------------- | -------- | ----------------------------------------------------------------------------------------------------------------- |
| id            | Path<T>                | no       | Unique name of this component.                                                                                    |
| labelText     | String                 | no       | Short description that will be displayed above the input field if set.                                            |
| isArea        | boolean                | yes      | If true, multi line text input will be enabled (default false).                                                   |
| required      | boolean                | yes      | If true, the user will not be able to submit the form when no input was provided (default false).                 |
| type          | HTMLInputTypeAttribute | yes      | The type of input that should be accepted by this input field (number, text, time, etc. default:'text')           |
| min           | number                 | yes      | Only applies if _type_ is "number". The Input must be greater than or equal to this value.                        |
| max           | number                 | yes      | Only applies if _type_ is "number". The Input must be less than or equal to this value.                           |
| register      | UseFormRegister<T>     | yes      | React hook form register. See https://www.react-hook-form.com/api/useform/register/ for details.                  |
| onChange      | () => void             | yes      | Callback that is invoked if the user changes the text field (e.g. after adding a new character).                  |
| valueAsNumber | boolean                | yes      | Only applies if _type_ is "number". Indicates that the value should be returned as a number type. (default false) |
| errorTitle    | string                 | yes      | Will be applied to the html title attribute of the textfield.                                                     |
