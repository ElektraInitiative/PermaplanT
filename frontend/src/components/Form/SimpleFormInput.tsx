import { FieldValues, Path, UseFormRegister } from 'react-hook-form';

interface SimpleFormInputProps<T extends FieldValues>
  extends React.InputHTMLAttributes<HTMLInputElement> {
  id: Path<T>;
  labelText: string;
  register?: UseFormRegister<T>;
  valueAsNumber?: boolean;
  errorTitle?: string;
}

export default function SimpleFormInput<T extends FieldValues>({
  id,
  labelText = '',
  register,
  valueAsNumber = false,
  errorTitle,
  ...props
}: SimpleFormInputProps<T>) {
  // Extract the input fields value from the respective Events before calling onChange.
  /*  const callOnChange = function <E>(event: React.ChangeEvent<E> | React.KeyboardEvent<E>) {
    if (onChange == null) return;

    // If somebody finds a way of fixing the next line, please do!
    const value = (event.target as unknown as HTMLInputElement).value;

    if (type === 'number') {
      onChange(parseInt(value));
    }

    onChange(value);
  }; */

  return (
    <div className="dark:text-white">
      <label htmlFor={id} className="mb-2 block text-sm font-medium">
        {labelText}
        {props.required ? <span className="text-red-800"> *</span> : <></>}
      </label>
      <input
        id={id}
        title={errorTitle}
        {...props}
        {...register?.(id, {
          valueAsNumber: props.type === 'number' || valueAsNumber,
        })}
        className="block h-11 w-full rounded-lg border border-neutral-500 bg-neutral-100 p-2.5 text-sm placeholder-neutral-500 focus:border-primary-500 focus:outline-none dark:border-neutral-400-dark dark:bg-neutral-50-dark dark:focus:border-primary-300"
      />
    </div>
  );
}
