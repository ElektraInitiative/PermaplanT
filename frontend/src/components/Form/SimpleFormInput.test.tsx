import SimpleFormInput from './SimpleFormInput';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';

it('renders correctly', () => {
  const tree = renderer
    .create(
      <MemoryRouter>
        <SimpleFormInput id="test" labelText="Some Label" />
      </MemoryRouter>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
