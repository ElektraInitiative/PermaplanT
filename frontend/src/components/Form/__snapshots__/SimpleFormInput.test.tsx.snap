// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="dark:text-white"
>
  <label
    className="mb-2 block text-sm font-medium"
    htmlFor="test"
  >
    Some Label
  </label>
  <input
    className="block h-11 w-full rounded-lg border border-neutral-500 bg-neutral-100 p-2.5 text-sm placeholder-neutral-500 focus:border-primary-500 focus:outline-none dark:border-neutral-400-dark dark:bg-neutral-50-dark dark:focus:border-primary-300"
    id="test"
  />
</div>
`;
